File: Python/BackTracking/18_partition_array_to_k_subsets.py
"""
Idea:
calculating target sum of each subset, and using backtracking to find , if there exist any combination
where each subset sums to target value.
"""


class Solution:
    def isKPartitionPossible(self, a, k):
        # Calculate the total sum of the array elements
        s = sum(a)

        # If the sum is not divisible evenly by k, return False
        if s % k != 0:
            return False

        # Calculate the target sum for each subset
        t = s // k

        # Initialize an array to track the sum of elements in each subset
        subset = [0] * k

        # Recursive function to check if it's possible to partition the array
        def fun(idx, n, subset, t):
            # If all elements are processed, return True
            if idx == n:
                return True

            # Try to add the current element to each subset
            for i in range(k):
                if subset[i] + a[idx] <= t:
                    subset[i] += a[idx]

                    # Recursively check the next element
                    if fun(idx + 1, n, subset, t):
                        return True

                    # Backtrack: undo the addition if it doesn't lead to a solution
                    subset[i] -= a[idx]

                    # Optimization: if the subset sum is 0, no need to try further
                    if subset[i] == 0:
                        break

            # If no subset is found, return False
            return False

        # Start the recursive function with initial parameters
        return fun(0, len(a), subset, t)


# Driver Code Starts


if __name__ == '__main__':
    tcs = int(input())
    for _ in range(tcs):
        N = int(input())
        arr = [int(x) for x in input().split()]
        k = int(input())
        if Solution().isKPartitionPossible(arr, k):
            print(1)
        else:
            print(0)
# } Driver Code Ends

File: C++/BackTracking/18_partition_array_to_k_subsets.cpp
/*
Idea:
calculating target sum of each subset, and using backtracking to find , if there exist any combination
where each subset sums to target value.
*/

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
/*You are required to complete this method */

class Solution
{
public:
    // Function to check if it's possible to partition the array into k subsets with equal sum
    bool isKPartitionPossible(int a[], int n, int k)
    {
        // Calculate the total sum of the array elements
        int s = 0;
        for (int i = 0;i < n;++i)
            s += a[i];


        // If the sum is not divisible evenly by k, return false
        if (s % k != 0)
            return false;


        // Calculate the target sum for each subset
        int t = s / k;

        // Initialize an array to track the sum of elements in each subset
        vector<int> subset(k, 0);

        // Start the recursive function with initial parameters
        return fun(a, 0, n, subset, t);
    }

private:
    // Recursive function to explore different possibilities of partitioning
    bool fun(int a[], int idx, int n, vector<int>& subset, int t)
    {
        // If all elements are processed, return true
        if (idx == n)
            return true;


        // Try to add the current element to each subset
        for (int i = 0; i < subset.size(); ++i)
        {
            if (subset[i] + a[idx] <= t)
            {
                subset[i] += a[idx];

                // Recursively check the next element
                if (fun(a, idx + 1, n, subset, t))
                    return true;


                // Backtrack: undo the addition if it doesn't lead to a solution
                subset[i] -= a[idx];

                // Optimization: if the subset sum is 0, no need to try further
                if (subset[i] == 0)
                    break;

            }
        }

        // If no subset is found, return false
        return false;
    }
};


//{ Driver Code Starts.
int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n;

        cin >> n;
        int a[n];
        for (int i = 0;i < n;i++)
            cin >> a[i];
        int k;
        cin >> k;
        Solution obj;
        cout << obj.isKPartitionPossible(a, n, k) << endl;
    }
}
// } Driver Code Ends